src/hyperlink/_url.py

```diff
@@ -749,37 +775,46 @@ class URL(object):
-          the default port of the scheme, if it is known. See the
-          ``SCHEME_PORT_MAP`` and :func:`register_default_port`
-          for more info.
-       path (tuple): A tuple of strings representing the
-          slash-separated parts of the path.
-       query (tuple): The query parameters, as a dictionary or
-          as an iterable of key-value pairs.
-       fragment (unicode): The fragment part of the URL.
-       rooted (bool): Whether or not the path begins with a slash.
-       userinfo (unicode): The username or colon-separated
-          username:password pair.
-       uses_netloc (bool): Indicates whether two slashes appear
-          between the scheme and the host (``http://eg.com`` vs
-          ``mailto:e@g.com``). Set automatically based on scheme.
-
-    All of these parts are also exposed as read-only attributes of
-    URL instances, along with several useful methods.
+        scheme: The text name of the scheme.
+        host: The host portion of the network location
+        port: The port part of the network location. If ``None`` or no port is
+            passed, the port will default to the default port of the scheme, if
+            it is known. See the ``SCHEME_PORT_MAP`` and
+            :func:`register_default_port` for more info.
+        path: A tuple of strings representing the slash-separated parts of the
+            path.
+        query: The query parameters, as a dictionary or as an iterable of
+            key-value pairs.
+        fragment: The fragment part of the URL.
+        rooted: Whether or not the path begins with a slash.
+        userinfo: The username or colon-separated username:password pair.
+        uses_netloc: Indicates whether two slashes appear between the scheme
+            and the host (``http://eg.com`` vs. ``mailto:e@g.com``).
+            Set automatically based on scheme.
+
+    All of these parts are also exposed as read-only attributes of URL
+    instances, along with several useful methods.
 
     .. _RFC 3986: https://tools.ietf.org/html/rfc3986
     .. _RFC 3987: https://tools.ietf.org/html/rfc3987
     """  # noqa: E501
 
-    def __init__(self, scheme=None, host=None, path=(), query=(), fragment=u'',
-                 port=None, rooted=None, userinfo=u'', uses_netloc=None):
+    def __init__(
+        self,
+        scheme=None,       # type: Optional[Text]
+        host=None,         # type: Optional[Text]
+        path=(),           # type: Iterable[Text]
+        query=(),          # type: QueryParameter
+        fragment=u"",      # type: Text
+        port=None,         # type: Optional[int]
+        rooted=None,       # type: Optional[bool]
+        userinfo=u"",      # type: Text
+        uses_netloc=None,  # type: Optional[bool]
+    ):
+        # type: (...) -> None
         if host is not None and scheme is None:
             scheme = u'http'  # TODO: why
-        if port is None:
+        if port is None and scheme is not None:
```

Comment:
The added check is because the following call to `.get` expects `Text`, not `Optional[Text]`, and `None` isn't in the dict, so the result would be `None`, which is already what `port` is set to here, so there you go.